### PR TITLE
Fix Nuget CUDA Packaging pipeline

### PR DESCRIPTION
**Pull request by jchen351**

### CODE REVIEW

File: tools/ci_build/github/azure-pipelines/stages/nuget-combine-cuda-stage.yml [hunk 1]
```diff
@@ -189,8 +189,25 @@ stages:
           parameters:
             PackageType: 'nuget'
             PackagePath: '$(Build.ArtifactStagingDirectory)'
-            PackageName: 'Microsoft.ML.OnnxRuntime.*nupkg'
             PlatformsSupported: 'win-x64,linux-x64'
+            # 1* stands for version number. we use it to fileter Gpu.Windows and Gpu.Linux packages
```
Comment (snnn) [inline tools/ci_build/github/azure-pipelines/stages/nuget-combine-cuda-stage.yml:193]: typo: `fileter`

File: tools/ci_build/github/azure-pipelines/stages/nuget-combine-cuda-stage.yml [hunk 2]
```diff
@@ -189,8 +189,25 @@ stages:
...
+            PackageName: 'Microsoft.ML.OnnxRuntime.Gpu.1*nupkg'
```
Comment (HectorSVC) [inline tools/ci_build/github/azure-pipelines/stages/nuget-combine-cuda-stage.yml:194]: >1* [](start = 55, length = 2)

how does 1* work? does it still work for string like Microsoft.ML.OnnxRuntime.Gpu.2.0.0.dev.xxx.nupkg?

Comment (snnn) [reply tools/ci_build/github/azure-pipelines/stages/nuget-combine-cuda-stage.yml:194]: I believe it would not.

Comment (HectorSVC) [reply tools/ci_build/github/azure-pipelines/stages/nuget-combine-cuda-stage.yml:194]: also found this
tools/nuget/validate_package.py:    is_gpu_package = bool("microsoft.ml.onnxruntime.gpu.1" in args.package_name.lower())

---
In reply to: [1446715440](https://github.com/microsoft/onnxruntime/pull/19054#discussion_r1446715440) [](ancestors = 1446715440)

Comment (jchen351) [reply tools/ci_build/github/azure-pipelines/stages/nuget-combine-cuda-stage.yml:194]: Let's fix it in a different PR. I copied it from existing implementations from Yi Zhang's PR. If we want to change it, we need to fix both places.

**Review by HectorSVC — APPROVED** [review]

**Comment by snnn** [comment]
The pipeline still fails:

Microsoft.ML.OnnxRuntime.EndToEndTests\Microsoft.ML.OnnxRuntime.EndToEndTests.csproj : warning NU1603: Microsoft.ML.OnnxRuntime.Gpu 1.17.0-dev-20240110-0114-1345563c79 depends on Microsoft.ML.OnnxRuntime.Gpu.Linux (>= 1.17.0-dev-20240110-0114-1345563c79) but Microsoft.ML.OnnxRuntime.Gpu.Linux 1.17.0-dev-20240110-0114-1345563c79 was not found. An approximate best match of Microsoft.ML.OnnxRuntime.Gpu.Linux 1.17.0-dev-20240110-0115-1345563c79 was resolved.
Microsoft.ML.OnnxRuntime.EndToEndTests\Microsoft.ML.OnnxRuntime.EndToEndTests.csproj : warning NU1603: Microsoft.ML.OnnxRuntime.Gpu 1.17.0-dev-20240110-0114-1345563c79 depends on Microsoft.ML.OnnxRuntime.Gpu.Windows (>= 1.17.0-dev-20240110-0114-1345563c79) but Microsoft.ML.OnnxRuntime.Gpu.Windows 1.17.0-dev-20240110-0114-1345563c79 was not found. An approximate best match of Microsoft.ML.OnnxRuntime.Gpu.Windows 1.17.0-dev-20240110-0115-1345563c79 was resolved.
Microsoft.ML.OnnxRuntime.EndToEndTests\Microsoft.ML.OnnxRuntime.EndToEndTests.csproj : error NU1605: Warning As Error: Detected package downgrade: Microsoft.ML.OnnxRuntime.Managed from 1.17.0-dev-20240110-0115-1345563c79 to 1.17.0-dev-20240110-0114-1345563c79. Reference the package directly from the project to select a different version. 
Microsoft.ML.OnnxRuntime.EndToEndTests\Microsoft.ML.OnnxRuntime.EndToEndTests.csproj : error NU1605:  Microsoft.ML.OnnxRuntime.EndToEndTests -> Microsoft.ML.OnnxRuntime.Gpu 1.17.0-dev-20240110-0114-1345563c79 -> Microsoft.ML.OnnxRuntime.Gpu.Windows 1.17.0-dev-20240110-0115-1345563c79 -> Microsoft.ML.OnnxRuntime.Managed (>= 1.17.0-dev-20240110-0115-1345563c79) 
Microsoft.ML.OnnxRuntime.EndToEndTests\Microsoft.ML.OnnxRuntime.EndToEndTests.csproj : error NU1605:  Microsoft.ML.OnnxRuntime.EndToEndTests -> Microsoft.ML.OnnxRuntime.Gpu 1.17.0-dev-20240110-0114-1345563c79 -> Microsoft.ML.OnnxRuntime.Managed (>= 1.17.0-dev-20240110-0114-1345563c79)
Microsoft.ML.OnnxRuntime.EndToEndTests\Microsoft.ML.OnnxRuntime.EndToEndTests.csproj : error NU1605: Warning As Error: Detected package downgrade: Microsoft.ML.OnnxRuntime.Managed from 1.17.0-dev-20240110-0115-1345563c79 to 1.17.0-dev-20240110-0114-1345563c79. Reference the package directly from the project to select a different version. 
Microsoft.ML.OnnxRuntime.EndToEndTests\Microsoft.ML.OnnxRuntime.EndToEndTests.csproj : error NU1605:  Microsoft.ML.OnnxRuntime.EndToEndTests -> Microsoft.ML.OnnxRuntime.Gpu 1.17.0-dev-20240110-0114-1345563c79 -> Microsoft.ML.OnnxRuntime.Gpu.Linux 1.17.0-dev-20240110-0115-1345563c79 -> Microsoft.ML.OnnxRuntime.Managed (>= 1.17.0-dev-20240110-0115-1345563c79) 
Microsoft.ML.OnnxRuntime.EndToEndTests\Microsoft.ML.OnnxRuntime.EndToEndTests.csproj : error NU1605:  Microsoft.ML.OnnxRuntime.EndToEndTests -> Microsoft.ML.OnnxRuntime.Gpu 1.17.0-dev-20240110-0114-1345563c79 -> Microsoft.ML.OnnxRuntime.Managed (>= 1.17.0-dev-20240110-0114-1345563c79)

**Comment by jchen351** [comment]
Looks like Windows package is also depending on the Linux one.

**Comment by mszhanyi** [comment]
> Looks like Windows package is also depending on the Linux one.

@jchen351, did you add -p:CurrentTime=$(BuildTime) as
https://github.com/microsoft/onnxruntime/blob/fd6bab4250c41a7f6498e6fa02ba446bc74e0a8d/tools/ci_build/github/azure-pipelines/c-api-noopenmp-packaging-pipelines.yml#L851

**Comment by mszhanyi** [comment]
The usage is here
https://github.com/microsoft/onnxruntime/blob/fd6bab4250c41a7f6498e6fa02ba446bc74e0a8d/csharp/OnnxRuntime.CSharp.proj#L51-L54

**Comment by jchen351** [comment]
> > Looks like Windows package is also depending on the Linux one.
> 
> @jchen351, did you add -p:CurrentTime=$(BuildTime) as
> 
> https://github.com/microsoft/onnxruntime/blob/fd6bab4250c41a7f6498e6fa02ba446bc74e0a8d/tools/ci_build/github/azure-pipelines/c-api-noopenmp-packaging-pipelines.yml#L851

No, But I merge your 2 new E2E testing stages back to original GPU e2e test. See commit [628fe28](https://github.com/microsoft/onnxruntime/pull/19054/commits/628fe282a369d28ef3af1cbe13bfe57c82443970)

**Comment by snnn** [comment]
@mszhanyi , going forward can we avoid taking dependency on date/time ? We may use the pipeline's buildid instead. Because in the same pipeline run the date/time might be different. The starting time of each job is different.

**Comment by skottmckay** [comment]
> @mszhanyi Yi Zhang FTE , going forward can we avoid taking dependency on date/time ? We may use the pipeline's buildid instead. Because in the same pipeline run the date/time might be different. The starting time of each job is different. So it's hard to make connections to the jobs.

Date/time is much better for context. e.g. if there's only a build id I would need to go back to the CI to try and work out which commit was involved and what date/time the package was created. Very indirect and I also need to know the CI that produced the package - which can be non-trivial to discover.

I would expect it's a one-off to make sure all the CI builds for one overall set of packages specify a consistent date/time value when creating the C# packages. Given that, I'd strongly prefer we keep date/time if possible.

**Review by snnn — APPROVED** [review]

**Pull request merged**